libraries/Microsoft.Bot.Builder.Azure/CosmosDbSqlStorage.cs

```diff
@@ -0,0 +1,195 @@
+        /// <param name="authKey">The AuthKey used by the client from the Azure Cosmos DB service.</param>
+        /// <param name="databaseId">The Database ID.</param>
+        /// <param name="collectionId">The Collection ID.</param>
+        public CosmosDbSqlStorage(Uri serviceEndpoint, string authKey, string databaseId, string collectionId)
+        {
+            if (serviceEndpoint == null)
+            {
+                throw new ArgumentNullException(nameof(serviceEndpoint), "Service EndPoint for CosmosDB is required.");
+            }
+
+            if (string.IsNullOrEmpty(authKey))
+            {
+                throw new ArgumentException("AuthKey for CosmosDB is required.", nameof(authKey));
+            }
+
+            if (string.IsNullOrEmpty(databaseId))
+            {
+                throw new ArgumentException("DatabaseId is required.", nameof(databaseId));
+            }
+
+            if (string.IsNullOrEmpty(collectionId))
+            {
+                throw new ArgumentException("CollectionId is required.", nameof(collectionId));
+            }
+
+            this.databaseId = databaseId;
+            this.collectionId = collectionId;
+
+            this.client = new DocumentClient(serviceEndpoint, authKey, connectionPolicy);
+            this.client.CreateDatabaseIfNotExistsAsync(new Database { Id = databaseId })
+                .Wait();
+
+            this.client.CreateDocumentCollectionIfNotExistsAsync(UriFactory.CreateDatabaseUri(databaseId), new DocumentCollection { Id = collectionId })
+                .Wait();
+        }
+
+        /// <summary>
+        /// Deletes the specified keys from storage.
+        /// </summary>
+        /// <param name="keys">List of keys to remove.</param>
+        /// <returns></returns>
+        public async Task Delete(params string[] keys)
+        {
+            var tasks = keys.Select(key =>
+                this.client.DeleteDocumentAsync(
+                    UriFactory.CreateDocumentUri(this.databaseId, this.collectionId, SanitizeKey(key))));
+
+            await Task.WhenAll(tasks).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Returns an instance of <see cref="StoreItems"/> bag with the values for the specified  keys.
+        /// </summary>
+        /// <param name="keys">List of keys to retrieve.</param>
+        /// <returns>A <see cref="StoreItems"/> instance.</returns>
+        public async Task<StoreItems> Read(params string[] keys)
+        {
+            var storeItems = new StoreItems();
+            foreach (string key in keys)
+            {
+                try
+                {
+                    var uri = UriFactory.CreateDocumentUri(this.databaseId, this.collectionId, SanitizeKey(key));
+                    var response = (await this.client.ReadDocumentAsync<DocumentStoreItem>(uri).ConfigureAwait(false));
```

Comment:
Improvement: Instead of querying key by key, create a SQL Query using WHERE IN. E.g.:
`
SELECT * FROM c
WHERE c.id in ('key1', key2')
`

Reply:
Code updated to use DocumentQuery and ExecuteNextAsync to retrieve all key-items